### PR TITLE
[DOC] Improvement of documentation

### PR DESCRIPTION
**Pull request by Mildophin**
This MR adds a folder named 'examples' that is filled with python code snippets representing parts of code we want to display to the users when they start using pasqal-cloud and send their first batch of jobs.

The updated code snippet in the readme is not visible in this MR because the link has to target a commit or branch name and I put the dev branch to makes sure it doesn't break when I will close this MR and also to have the snippet displayed updated automatically (we still have to change the last line number if we add or remove code).

Also, this MR is to have only one source of truth for all the tutorials linked to pasqal-cloud, so we only have to embed the code of these examples instead of updating each tutorial one by one.

Edit: https://github.com/pasqal-io/pasqal-cloud/pull/104#issuecomment-1740932760

### CODE REVIEW

File: examples/send_batch_with_jobs.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+from pasqal_cloud import SDK
+from pasqal_cloud.device import EmulatorType
+
+project_id = (
+    "your_project_id"  # Replace this value by your project_id on the PASQAL platform.
```
Comment (oliver-gordon) [inline examples/send_batch_with_jobs.py:5]: `# Replace this value by ` -> `# Replace this value with`

File: examples/send_batch_with_jobs.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+)
+username = "your_username"  # Replace this value by your username or email on the PASQAL platform.
```
Comment (oliver-gordon) [inline examples/send_batch_with_jobs.py:7]: `# Replace this value by ` -> `# Replace this value with`

File: examples/send_batch_with_jobs.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+password = (
+    "your_password"  # Replace this value by your password on the PASQAL platform.
```
Comment (oliver-gordon) [inline examples/send_batch_with_jobs.py:9]: `# Replace this value by ` -> `# Replace this value with`

**Comment by awennersteen** [comment]
I like the idea!

Would it be much effort to put the rest of the code snippets in the README into the same format?

Are you sure it wouldn't be possible to make the links relative somehow?
We do have a master branch (although we don't use it much :( ) for example.

File: README.md [hunk 4]
```diff
@@ -51,36 +51,9 @@ Once you have created your sequence, you should serialize it as follows:
...
-# For using a basic single-threaded QPU emulator that can go up to 10 qubits, you can specify the "EMU_FREE" emulator.
-from pasqal_cloud.device import EmulatorType
-batch = sdk.create_batch(serialized_sequence, [job1,job2], emulator=EmulatorType.EMU_FREE)
-
-# Once the QPU has returned the results, you can access them with the following:
-for job in batch.ordered_jobs:
-    print(job.result)
-
-```
+https://github.com/pasqal-io/pasqal-cloud/blob/dev/examples/send_batch_with_jobs.py#L1-L28
```
Comment (oliver-gordon) [inline README.md:56]: Did you check out what it looks like with multiple, smaller scoped embedding rather than 1 large embedded of 28 lines? If you were not content with the scrolling effect

Is this the same file that will embed on docusaurus?

Comment (Mildophin) [reply README.md:56]: Tbh no I didn't as I don't think it will look nice as at the bottom of each snippet there is a footer with the path of the file.

Yes it will be the same, plus the file of the generation of the pulser sequence I added in this branch.

File: examples/send_batch_with_jobs.py (hunk 3 above)
Comment (Mildophin) [reply examples/send_batch_with_jobs.py:9]: I will update this error in the others parts of the readme, thx for spotting it

**Comment by Mildophin** [comment]
> I like the idea!
> 
> Would it be much effort to put the rest of the code snippets in the README into the same format?
> 
> Are you sure it wouldn't be possible to make the links relative somehow? We do have a master branch (although we don't use it much :( ) for example.
@awennersteen 

It's really hard, tbh i'm having headaches of how difficult it is to embed python code from another file in markdown files, I'm currently thinking they are trying to prevent it for a reason. The best solution I see for now is switching to readthedoc (sphinx) to embed snippets properly. Exactly like they did for Pulser.

For the branch master, well, not used, so not considered haha

File: README.md (hunk 4 above)
Comment (Mildophin) [reply README.md:56]: An alternative being to display a link to the code snippet instead.. but is it okay ? we lose all the visual aspect and it adds an effort for the users

**Comment by Mildophin** [comment]
After giving it enough thought, I am dropping the idea of a unique source of truth for now. I think it will be more useful later when we will build a more detailed documentation for pasqal cloud, for now I think that the simplest way for developers to update the documentation is to do it by hand.

I will change this MR to only update the code examples in the readme. Then I will set up a checklist for each MR on pasqal cloud, telling everyone what to check before merging it but also what to do after it is merged.

**Comment by Mildophin** [comment]
To help you reviewing it, here is the link to the readme file: https://github.com/pasqal-io/pasqal-cloud/blob/5829093004f4545bc826559bde327bd1f821204d/README.md

**Comment by Mildophin** [comment]
I added a pull request in this MR to make sure we don't forget any point while working on pasqal-cloud. Feel free to correct me if I forgot something or if my english is broken.

File: .github/pull_request_template.md [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+<!-- Here, add all the breaking changes that this MR involves if there is any -->
+
+### Checklist
+
+- [ ] Title of the MR follows the right format (see the internal documentation).
+
+#### Documentation
+
+- [ ] Update the version of pasqal-cloud in `_version.py` following the changes your MR implies. We use an **x**.**y**.**z** versioning, where **x** is not used yet, **y** for a breaking change and **z** for a new feature, bug correction or minor change.
```
[the remark this answers is not in the document]
Comment (Mildophin) [reply .github/pull_request_template.md:27]: Good point

File: README.md [hunk 6]
```diff
@@ -37,111 +37,126 @@ pre-commit install
...
+    print(f"job-id: {job.id}, status: {job.status}, result: {job.result}")
+```
 
-project_id="your_project_id" # Replace this value by your project_id. It can be found on the user portal under the name "group_id". "Groups" have recently been renamed to "projects" and frontend is currently being updated accordingly.
+### Create a workload
 
-username="your_username" # Replace this value by your username or email on the PASQAL platform.
-password="your_password" # Replace this value by your password on the PASQAL platform.
-# Ideally, do not write this password in a script but provide in through the command-line or as a secret environment variable.
+You can create a workload through the SDK with the following command:
```
[the remark this answers is not in the document]
Comment (Mildophin) [reply README.md:134]: yes some explanation could be great, is there any doc already somewhere that I could reuse?

File: .github/pull_request_template.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+<!-- Here, add any extra criteria you consider mandatory for merging on top of the usual criteria -->
+
+#### Breaking changes
+
+<!-- Here, add all the breaking changes that this MR involves if there is any -->
+
+### Checklist
+
+- [ ] Title of the MR follows the right format (see the internal documentation).
```
Comment (oliver-gordon) [inline .github/pull_request_template.md:23]: Given this is a public repo, maybe provide instruction for the format somewhere here? See internal documentation not very helpful

File: .github/pull_request_template.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (oliver-gordon) [reply .github/pull_request_template.md:27]: `the changes your MR implies` reads a little strange.  'following the changes in your PR` perhaps.

We should reference PR not MR on Github since they call them pull requests

File: .github/pull_request_template.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+- [ ] Update the version of pasqal-cloud in `_version.py` following the changes your MR implies. We use an **x**.**y**.**z** versioning, where **x** is not used yet, **y** for a breaking change and **z** for a new feature, bug correction or minor change.
+- [ ] Update CHANGELOG.md with a description explaining briefly the changes to the users without using technical vocabulary.
+
+#### Tests
+
+- [ ] Unit tests have been added or adjusted.
+- [ ] Local setup was done & tests were run locally.
+
+#### E2E tests
+- [ ] In case your MR changes the current logic, update the E2E tests and then launch them while targeting the branch of this MR.
```
Comment (oliver-gordon) [inline .github/pull_request_template.md:36]: `MR` -> `PR`

File: .github/pull_request_template.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+- [ ] Update the version of pasqal-cloud in `_version.py` following the changes your MR implies. We use an **x**.**y**.**z** versioning, where **x** is not used yet, **y** for a breaking change and **z** for a new feature, bug correction or minor change.
+- [ ] Update CHANGELOG.md with a description explaining briefly the changes to the users without using technical vocabulary.
+
+#### Tests
+
+- [ ] Unit tests have been added or adjusted.
+- [ ] Local setup was done & tests were run locally.
+
+#### E2E tests
```
Comment (oliver-gordon) [inline .github/pull_request_template.md:35]: Should this be here? It's a public repo so does not make sense to have this information anywhere but internally

File: .github/pull_request_template.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+- [ ] In case your MR changes the current logic, update the E2E tests and then launch them while targeting the branch of this MR.
+If it is not changing the current code functioning, test the branch of your MR with the E2E to make sure nothing is broken.
+
+#### After merging
+
+- [ ] Open an MR on the E2E that updates the version of the sdk used in dependencies.
```
Comment (oliver-gordon) [inline .github/pull_request_template.md:41]: Same issue here

**Review by oliver-gordon — CHANGES_REQUESTED** [review]
We shouldn't reference internal things in a public doc, CC internal conversation to take place

File: .github/pull_request_template.md [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+#### Documentation
+
+- [ ] Update the version of pasqal-cloud in `_version.py` following the changes your MR implies. We use an **x**.**y**.**z** versioning, where **x** is not used yet, **y** for a breaking change and **z** for a new feature, bug correction or minor change.
+- [ ] Update CHANGELOG.md with a description explaining briefly the changes to the users without using technical vocabulary.
+
+#### Tests
+
+- [ ] Unit tests have been added or adjusted.
+- [ ] Local setup was done & tests were run locally.
```
Comment (oliver-gordon) [inline .github/pull_request_template.md:33]: `Local setup was done & tests were run locally.` would be fine as `Tests were run locally`.

The first sentence is a little odd

File: .github/pull_request_template.md [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+<!-- Here, add all the breaking changes that this MR involves if there is any -->
+
+### Checklist
+
+- [ ] Title of the MR follows the right format (see the internal documentation).
+
+#### Documentation
+
+- [ ] Update the version of pasqal-cloud in `_version.py` following the changes your MR implies. We use an **x**.**y**.**z** versioning, where **x** is not used yet, **y** for a breaking change and **z** for a new feature, bug correction or minor change.
+- [ ] Update CHANGELOG.md with a description explaining briefly the changes to the users without using technical vocabulary.
```
Comment (oliver-gordon) [inline .github/pull_request_template.md:28]: `without using technical vocabulary.` <- why this part? I would assume anyone reading a change log on GitHub is most likely going to understand some technical vocabulary

File: .github/pull_request_template.md (hunk 9 above)
Comment (oliver-gordon) [inline .github/pull_request_template.md:35]: So I'd prefer it if we can reword some of these check list items to be neutral to internal/external developers, including changing internal references to something more comprehensible `E2E` might make sense to some but `Internal test pipeline` will make sense to everyone

File: .github/pull_request_template.md [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+#### Tests
+
+- [ ] Unit tests have been added or adjusted.
+- [ ] Local setup was done & tests were run locally.
+
+#### E2E tests
+- [ ] In case your MR changes the current logic, update the E2E tests and then launch them while targeting the branch of this MR.
+If it is not changing the current code functioning, test the branch of your MR with the E2E to make sure nothing is broken.
```
Comment (oliver-gordon) [inline .github/pull_request_template.md:37]: Personally I'd reword the sentence, it reads a little strange.
I'd opt for:
`If your PR hasn't changed any functionality, it still needs to be validated against internal tests`.

File: .github/pull_request_template.md (hunk 12 above)
Comment (Mildophin) [reply .github/pull_request_template.md:28]: The idea I meant was to not go too much in details

File: .github/pull_request_template.md (hunk 7 above)
Comment (Mildophin) [reply .github/pull_request_template.md:23]: Maybe having the document in accessible to public ? Isn't something that need to be internal right ?

File: README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (Mildophin) [reply README.md]: Added

**Comment by Mildophin** [comment]
Updated.

Link for the readme: https://github.com/pasqal-io/pasqal-cloud/blob/mv/improvement-of-documentation-examples/README.md

Link for the PR template: https://github.com/pasqal-io/pasqal-cloud/blob/mv/improvement-of-documentation-examples/.github/pull_request_template.md

File: README.md [hunk 14]
```diff
@@ -37,143 +37,160 @@ pre-commit install
...
 
-You can create a workload through the SDK with the following command:
 ```python
-workload=sdk.create_workload(workload_type="<WORKLOAD_TYPE>",backend="<BACKEND>",config={"config_param_1":"value"})
+serialized_sequence = sequence.to_abstract_repr()
+```
 
-#You can cancel the workload by doing:
-sdk.cancel_workload(workload.id)
+Once you are login and have serialized your sequence, the last step is to define your jobs.
```
Comment (oliver-gordon) [inline README.md:100]: Nit: This sentence doesn't make too much sense, we don't actually log in with the SDK and `once you are login` wouldn't be correct still

File: README.md [hunk 15]
```diff
@@ -37,143 +37,160 @@ pre-commit install
...
-The package main component is a python object called `SDK` which can be used to create a `Batch` and send it automatically
-to Pasqal APIs using an API token generated in the [user portal](https://portal.pasqal.cloud).
 
-A `Batch` is a group of jobs with the same sequence that will run on the same QPU. For each job of a given batch you must set a value for each variable, if any, defined in your sequence.
-The batch sequence can be generated using [Pulser](https://github.com/pasqal-io/Pulser). See their [documentation](https://pulser.readthedocs.io/en/stable/),
-for more information on how to install the library and create your own sequence.
+### Authentication
 
-Once you have created your sequence, you should serialize it as follows:
+There are several ways to provide a correct authentication using the SDK.
```
Comment (oliver-gordon) [inline README.md:43]: Suggestion: `There are several ways to provide a correct authentication using the SDK.` -> `There are several ways to authenticate using the SDK.`

File: README.md [hunk 16]
```diff
@@ -37,143 +37,160 @@ pre-commit install
...
-serialized_sequence = sequence.to_abstract_repr()
+from pasqal_cloud import SDK
+
+project_id = "your_project_id"  # Replace this value with your project_id on the PASQAL platform.
+username = "your_username"  # Replace this value with your username or email on the PASQAL platform.
+password = "your_password"  # Replace this value with your password on the PASQAL platform.
 ```
 
-Once you have serialized your sequence, you can send it with the SDK with the following code
+Ideally, do not write this password in a script but provide in through the command-line or as a secret environment variable.
```
Comment (oliver-gordon) [inline README.md:53]: Not sure we need this line, people will follow whatever practice they want and it's not particularly relevant

File: README.md [hunk 17]
```diff
@@ -37,143 +37,160 @@ pre-commit install
...
+sdk = SDK(username=username, project_id=project_id)
+```
 
-# You can also choose to run your batch on an emulator using the optional argument 'emulator'
-# For using a basic single-threaded QPU emulator that can go up to 10 qubits, you can specify the "EMU_FREE" emulator.
-from pasqal_cloud.device import EmulatorType
-batch = sdk.create_batch(serialized_sequence, [job1,job2], emulator=EmulatorType.EMU_FREE)
+#### Method 3: Use a custom token provider
+You can define a custom class to provide the token.
+For example, if you know your token, you can use that token to authenticate directly to our APIs as follows.
```
Comment (oliver-gordon) [inline README.md:72]: `if you know your token, you can use that token to authenticate directly to our APIs as follows.` -> `if you have a token, you can use it to authenticate with our APIs.`

File: README.md [hunk 18]
```diff
@@ -37,143 +37,160 @@ pre-commit install
...
+### Create a batch of jobs
+
+The package main component is a python object called `SDK` which can be used to create a `Batch` and send it automatically
+to Pasqal APIs using an API token generated in the [user portal](https://portal.pasqal.cloud).
+
+A `Batch` is a group of jobs with the same sequence that will run on the same QPU. For each job of a given batch, you must set a value for each variable, if any, defined in your sequence.
+The batch sequence can be generated using [Pulser](https://github.com/pasqal-io/Pulser). See their [documentation](https://pulser.readthedocs.io/en/stable/),
+for more information on how to install the library and create your own sequence.
+
+The sequence should be a pulser Sequence object, once you have created it, you should serialize it as follows:
```
Comment (oliver-gordon) [inline README.md:94]: Nit: `once you have created it` -> `once it's created, you can serialize like so`

File: README.md [hunk 19]
```diff
@@ -37,143 +37,160 @@ pre-commit install
...
+sdk = SDK(token_provider=CustomTokenProvider(), project_id=project_id)
 ```
+Alternatively, create a custom TokenProvider that inherits from ExpiringTokenProvider. You should define a
+custom _query_token method which fetches your token. See Auth0TokenProvider implementation for an example.
 
-### Workload Creation
+### Create a batch of jobs
+
+The package main component is a python object called `SDK` which can be used to create a `Batch` and send it automatically
+to Pasqal APIs using an API token generated in the [user portal](https://portal.pasqal.cloud).
```
Comment (MatthieuMoreau0) [inline README.md:88]: I'm not sure what this API token refers to. This is reminiscent of API keys but these have been deleted a year ago. I would suggest removing the reference to this token, unless I forgot something

File: README.md (hunk 14 above)
Comment (MatthieuMoreau0) [inline README.md:100]: nit: logged in

File: README.md [hunk 20]
```diff
@@ -37,143 +37,160 @@ pre-commit install
...
+```python
+class CustomTokenProvider(TokenProvider):
+    def get_token(self):
+        return "your-token" # Replace this value with your token
 
+
+sdk = SDK(token_provider=CustomTokenProvider(), project_id=project_id)
 ```
+Alternatively, create a custom TokenProvider that inherits from ExpiringTokenProvider. You should define a
+custom _query_token method which fetches your token. See Auth0TokenProvider implementation for an example.
```
Comment (MatthieuMoreau0) [inline README.md:83]: this is code: `Auth0TokenProvider`

**Review by MatthieuMoreau0 — COMMENTED** [review]
Overall this looks ok. I personally prefer having a single code block that I can copy paste at once while this breaks instructions down into many smaller code blocks. It's a matter of taste so no big deal. We maybe could provide fully fledged-out examples that can be run out of the box in a future PR

File: README.md [hunk 21]
```diff
@@ -37,143 +37,160 @@ pre-commit install
...
 
-There are several ways to provide a correct authentication using the SDK.
+```python
+from pasqal_cloud.device import EmulatorType
+batch = sdk.create_batch(
+    serialized_sequence, [job1, job2], emulator=EmulatorType.EMU_FREE
+)
+```
+
+Once the QPU has returned the results, you can access them with the following:
```
Comment (oliver-gordon) [inline README.md:125]: Nit: `Once the QPU has` -> `Once the API has returned`

File: README.md [hunk 22]
```diff
@@ -37,143 +37,160 @@ pre-commit install
...
+    print(f"job-id: {job.id}, status: {job.status}, result: {job.result}")
+```
 
-project_id="your_project_id" # Replace this value by your project_id. It can be found on the user portal under the name "group_id". "Groups" have recently been renamed to "projects" and frontend is currently being updated accordingly.
+### Create a workload
 
-username="your_username" # Replace this value by your username or email on the PASQAL platform.
-password="your_password" # Replace this value by your password on the PASQAL platform.
-# Ideally, do not write this password in a script but provide in through the command-line or as a secret environment variable.
+A workload is a single unit of work to be computed by emulators or qpu.
```
Comment (oliver-gordon) [inline README.md:134]: Nit: `qpu` -> `a QPU`

File: README.md [hunk 23]
```diff
@@ -37,143 +37,160 @@ pre-commit install
...
+It is defined by a type (qadence_circuit, algorithm_dqgm, …) and a config.
+This config must be generated with the appropriate library for the desired type.
```
Comment (oliver-gordon) [inline README.md:136]: Comment: This entire description reads quite vague, is this really the information we want to display?

File: .github/pull_request_template.md [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+
+<!-- Here, add all the breaking changes that this PR involves if there is any -->
+
+### Checklist
+
+- [ ] The title of the PR follows the right format: [{Label}] {Short Message}. Label examples: IMPROVEMENT, FIX, REFACTORING... Short message is about what your PR changes.
+
+#### Documentation
+
+- [ ] **[TEMPORARY]** — Update the version of pasqal-cloud in `_version.py` following the changes in your PR and by using [semantic versioning](https://semver.org/).
```
Comment (oliver-gordon) [inline .github/pull_request_template.md:27]: Question: Why temporary? Do we have a ticket to update this when it stops being?

File: .github/pull_request_template.md [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+- [ ] Update CHANGELOG.md with a description explaining briefly the changes to the users.
+
+#### Tests
+
+- [ ] Unit tests have been added or adjusted.
+- [ ] Tests were run locally.
+
+#### Internal tests pipeline (PASQAL developers only)
+- [ ] Update the internal tests then launch them while targeting the branch of this PR.
```
Comment (oliver-gordon) [inline .github/pull_request_template.md:36]: Nit: `Update the internal tests then launch them` -> `Update and run the internal tests while targeting...`

File: README.md (hunk 19 above)
Comment (awennersteen) [reply README.md:88]: Now you'll need the project ID though, so perhaps worth referencing that if not done elsewhere

File: README.md (hunk 23 above)
Comment (Mildophin) [reply README.md:136]: @MatthieuMoreau0 @awennersteen  Is it enought you think ?

File: .github/pull_request_template.md (hunk 24 above)
Comment (Mildophin) [reply .github/pull_request_template.md:27]: @anneclairelh If you can bring some explanation, I do not have them

File: README.md (hunk 19 above)
Comment (Mildophin) [reply README.md:88]: it is referenced above in the authentication part

File: README.md (hunk 14 above)
Comment (Mildophin) [reply README.md:100]: I removed it

File: README.md (hunk 20 above)
Comment (MatthieuMoreau0) [reply README.md:81]: This refers to a class in our code and does not read as English; hence it should be formatted as code, i.e. in back quotes like this `Auth0TokenProvider`. Same remark applies to `TokenProvider`, `ExpiringTokenProvider` and  `_query_token`

**Comment by Mildophin** [comment]
@MatthieuMoreau0  I do agree it is probably easier for users when it is in one big block but after discussing with @oliver-gordon I think that it is better this way as there are different ways of authenticating, we can target different emulators or the QPU, there are many options. Also explanations shouldn't be code comments. However I do agree later one it would be nice to implement a whole bloc of code that we can copy + a detailed documentation explaining all the possibilities pasqal-cloud can provide.

File: README.md (hunk 20 above)
Comment (Mildophin) [reply README.md:81]: Understood, should we just remove the last sentence in that case ?

File: README.md (hunk 23 above)
Comment (MatthieuMoreau0) [reply README.md:131]: A rewording suggestion:

```
A workload is a unit of work to be executed on Pasqal Cloud Services infrastucture.  To submit a new workload, select a type of workloads, target one of the available backends and provide a configuration object to execute it.
``` 

The last sentence is vague maybe we could tell the user to refer to user portal documentation for details on workload configurations. But unfortunately that page doesn't exist just yet..

File: README.md (hunk 20 above)
Comment (MatthieuMoreau0) [reply README.md:81]: No I don't think so, why do you want to do that ? 

I simply suggest making clear we refer to objects from the SDK. Similarly to how you would quote words in another language in a given text - just to make it clearer for the reader

Comment (Mildophin) [reply README.md:81]: Oh yes mb I have read wrongly

File: README.md (hunk 23 above)
Comment (Mildophin) [reply README.md:131]: I do agree it is a bit vague, should we list the current option and explain them very briefly ?

File: .github/pull_request_template.md (hunk 24 above)
Comment (MatthieuMoreau0) [reply .github/pull_request_template.md:27]: I also don't think we should write this as temporary, it is irrelevant to the contributor. The only thing that matters is whether it applies right now or not. [Documentation should be timeless.](https://developers.google.com/style/timeless-documentation)

Instead I would suggest indeed to create an issue to get rid of this if this is our goal

**Review by MatthieuMoreau0 — APPROVED** [review]

**Review by MatthieuMoreau0 — APPROVED** [review]

**Review by oliver-gordon — APPROVED** [review]
